infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/CorfuLogReplicationRuntime.java

```diff
@@ -0,0 +1,295 @@
+                ThreadFactoryBuilder().setNameFormat("runtime-fsm-worker").build());
+        this.communicationFSMConsumer = Executors.newSingleThreadExecutor(new
+                ThreadFactoryBuilder().setNameFormat("runtime-fsm-consumer").build());
+
+        initializeStates();
+        this.state = states.get(LogReplicationRuntimeStateType.WAITING_FOR_CONNECTIVITY);
+
+        log.info("Log Replication Runtime State Machine initialized");
+    }
+
+    /**
+     * Start Log Replication Communication FSM
+     */
+    public void start() {
+        // Start Consumer Thread for this state machine (dedicated thread for event consumption)
+        communicationFSMConsumer.submit(this::consume);
+        router.connect();
+    }
+
+    /**
+     * Initialize all states for the Log Replication Runtime FSM.
+     */
+    private void initializeStates() {
+        /*
+         * Log Replication Runtime State instances are kept in a map to be reused in transitions, avoid creating one
+         * per every transition (reduce GC cycles).
+         */
+        states.put(LogReplicationRuntimeStateType.WAITING_FOR_CONNECTIVITY, new WaitingForConnectionsState(this));
+        states.put(LogReplicationRuntimeStateType.VERIFYING_REMOTE_LEADER, new VerifyingRemoteLeaderState(this, communicationFSMWorkers, router));
+        states.put(LogReplicationRuntimeStateType.NEGOTIATING, new NegotiatingState(this, communicationFSMWorkers, router, metadataManager));
+        states.put(LogReplicationRuntimeStateType.REPLICATING, new ReplicatingState(this, sourceManager));
+        states.put(LogReplicationRuntimeStateType.STOPPED, new StoppedState(sourceManager));
+        states.put(LogReplicationRuntimeStateType.UNRECOVERABLE, new UnrecoverableState());
+    }
+
+    /**
+     * Input function of the FSM.
+     *
+     * This method enqueues runtime events for further processing.
+     *
+     * @param event LogReplicationRuntimeEvent to process.
+     */
+    public synchronized void input(LogReplicationRuntimeEvent event) {
+        try {
+            if (state.getType().equals(LogReplicationRuntimeStateType.STOPPED)) {
+                // Not accepting events, in stopped state
+                return;
+            }
+            eventQueue.put(event);
+        } catch (InterruptedException ex) {
+            log.error("Log Replication interrupted Exception: ", ex);
+        }
+    }
+
+    /**
+     * Consumer of the eventQueue.
+     * <p>
+     * This method consumes the log replication events and does the state transition.
+     */
+    private void consume() {
+        try {
+            if (state.getType() == LogReplicationRuntimeStateType.STOPPED) {
+                log.info("Log Replication Communication State Machine has been stopped. No more events will be processed.");
+                return;
+            }
+
+            //  Block until an event shows up in the queue.
+            LogReplicationRuntimeEvent event = eventQueue.take();
```

Comment:
There is a possible starving case that there are two standbys, standby1 call consume, but standby1's eventQueue is  empty, so the thread is blocking here even though standby2's eventQue is not empty.

Reply:
How can that happen if they don't share the same eventQueue nor run on the same thread? A CorfuLogReplicationRuntime is dedicated to a standby site.